In [30]:
import modules.singleton_reduction as singleton
import modules.isolate_reduction as isolate
from tqdm import tqdm
import importlib
importlib.reload(singleton)
importlib.reload(isolate)
import networkx as nx
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
from concurrent.futures import ThreadPoolExecutor
import io
import copy


In [23]:
# Load the full graph
edges = pd.read_csv("../data/edges_updated_reversed.csv", sep=' ')
edges_array = np.array(edges.loc[:, ['Source', 'Target']])
nodes = pd.read_csv(r"../data/nodes.csv", sep=",")
nodes_array = nodes["# index"]

G_full = nx.DiGraph()
G_full.add_edges_from(edges_array)
G_full.add_nodes_from(nodes_array)

In [103]:
importlib.reload(singleton)
importlib.reload(isolate)

G = copy.deepcopy(G_full)
G = singleton.singleton_reduction(G)
G_full_sr = copy.deepcopy(G)
# isolatedNodes = list(nx.isolates(G))
G, isolatedNum = isolate.isolate_reduction(G)

Node count before singleton reduction:  58742
Nodes before:  52124
Nodes after:  19257


In [104]:
import modules.sir_threaded as sir
import importlib
import copy
importlib.reload(sir)
init_infected = [9]
max_steps = 100
G_temp = copy.deepcopy(G)

# Uncomment below block if you want to only view nodes that are accessible from init_infected
G_temp = G_temp.subgraph(sir.get_accessible_sus_nodes(G_temp, init_infected)).copy()

In [ ]:
# Uncomment below block if you want to only view nodes that are accessible from init_infected
# print("Nodes to calculate: ", len(G_temp))
# pos = nx.spring_layout(G_temp, seed=8020, gravity=0.75)

In [105]:
importlib.reload(sir)

_, _, _, _, infotext_total, constants_total = sir.sir_model(G_temp, G_full_sr, init_infected=init_infected, infection_rate=0.2, recovery_rate=0.1, max_steps=max_steps, doSingletonReduction=True, noticeability_rates=(0.0001,0.01), network_type='full')

[DEBUG2] False
Running model...
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Step:  1
[DEBUG] QT:  0
[INFO] Infecting singletons and neighbors + Recovering infected nodes...
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG] Added 10 loss due to: Infection creation
[DEBUG

In [60]:
infotext_total

[{'ssw': 950,
  'sso': 5668,
  'it': 1,
  'st': 4990,
  'rt': 0,
  'qt': 0,
  'ss': 6618,
  'is': 0,
  'rs': 0},
 {'ssw': 801,
  'sso': 5668,
  'it': 617,
  'st': 4374,
  'rt': 0,
  'qt': 0,
  'ss': 6469,
  'is': 149,
  'rs': 0},
 {'ssw': 667,
  'sso': 5668,
  'it': 1148,
  'st': 3843,
  'rt': 0,
  'qt': 0,
  'ss': 6335,
  'is': 283,
  'rs': 0},
 {'ssw': 576,
  'sso': 5668,
  'it': 1580,
  'st': 3411,
  'rt': 0,
  'qt': 0,
  'ss': 6244,
  'is': 374,
  'rs': 0},
 {'ssw': 493,
  'sso': 5668,
  'it': 1955,
  'st': 3036,
  'rt': 0,
  'qt': 0,
  'ss': 6161,
  'is': 457,
  'rs': 0},
 {'ssw': 423,
  'sso': 5668,
  'it': 2300,
  'st': 2691,
  'rt': 0,
  'qt': 0,
  'ss': 6091,
  'is': 527,
  'rs': 0},
 {'ssw': 359,
  'sso': 5668,
  'it': 2607,
  'st': 2384,
  'rt': 0,
  'qt': 0,
  'ss': 6027,
  'is': 591,
  'rs': 0},
 {'ssw': 315,
  'sso': 5668,
  'it': 2895,
  'st': 2096,
  'rt': 0,
  'qt': 0,
  'ss': 5983,
  'is': 635,
  'rs': 0},
 {'ssw': 273,
  'sso': 5668,
  'it': 3102,
  'st': 1889,
  'rt

In [58]:
constants_total

{'tn': 58742, 'xt': 53751, 'virusFoundStep': 61}

In [108]:
# Graphical Analysis of a single run
plt.close()

# create data
x = range(len(infotext_total))
it_total = [i['it'] for i in infotext_total]
st_total = [i['st'] for i in infotext_total]
rt_total = [i['rt'] for i in infotext_total]
qt_total = [i['qt'] for i in infotext_total]

loss_total = [i['loss'] for i in infotext_total]
max_loss_total = max(loss_total)
loss_total = [i/max_loss_total*(constants_total['tn']-constants_total['xt']) for i in loss_total]


# plot lines
plt.plot(x, st_total, label = "Total susceptible", alpha=0.5, color='g')
plt.plot(x, it_total, label = "Total infected", alpha=0.5, color='r')
plt.plot(x, rt_total, label = "Total recovered", alpha=0.5, color='b')
plt.plot(x, qt_total, label = "Total quarantined", alpha=0.5, color='0.7')
plt.plot(x, loss_total, label = "Loss (scaled)", alpha=0.5, color='k')
plt.vlines(constants_total['virusFoundStep'], 0, constants_total['tn']-constants_total['xt'], color='k', linestyle='dotted', lw=1)
plt.legend()
plt.savefig(f"results/test.png", format="PNG", bbox_inches='tight')

In [95]:
# Loss analysis of a single run
plt.close()
loss = {'quarantine': 1, 'infected': 1}

x = range(len(infotext_total))
it_total = [i['it'] for i in infotext_total]
st_total = [i['st'] for i in infotext_total]
rt_total = [i['rt'] for i in infotext_total]
qt_total = [i['qt'] for i in infotext_total]

plt.stackplot(x, (
                    [i*loss['infected'] for i in it_total],
                    [i*loss['quarantine'] for i in qt_total]
                ),
             labels=['Loss from infected', 'Loss from quarantines'],
             colors=['#ff0000','#444444'],
              alpha=0.8)
plt.legend()
plt.savefig(f"results/test2.png", format="PNG", bbox_inches='tight')

In [141]:
importlib.reload(sir)
import os
from multiprocessing import Pool
from functools import partial

# Loss analysis of multiple runs
run_count = 1000

init_infected = [9]
# init_infected = 
max_steps = 1000
infection_rate = 0.2
recovery_rate = 0.1
noticeability_rates = (0.0001,0.01)
quarantine_length = 10
contact_tracing = 3
loss_matrix = ((1, 0),(0.2, 0.1))

loss_results = None

worker = partial(
    sir.simulate_threaded,
    G=G, 
    G_full_sr=G_full_sr, 
    init_infected=init_infected, 
    infection_rate=infection_rate, 
    recovery_rate=recovery_rate, 
    max_steps=max_steps, 
    noticeability_rates=noticeability_rates, 
    quarantine_length=quarantine_length,
    contact_tracing=contact_tracing,
    loss_matrix=loss_matrix,
)

with Pool(8) as p:
    loss_results = p.map(worker, range(run_count))

(np.mean(loss_results), np.std(loss_results))

Run:  0  --- Steps taken:  98  --- Loss:  13187.900000000007
Run:  32  --- Steps taken:  88  --- Loss:  15226.50000000001
Run:  64  --- Steps taken:  77  --- Loss:  7500.599999999999
Run:  1  --- Steps taken:  84  --- Loss:  11644.700000000003
Run:  33  --- Steps taken:  139  --- Loss:  17523.500000000036
Run:  96  --- Steps taken:  98  --- Loss:  17065.60000000002
Run:  65  --- Steps taken:  93  --- Loss:  10944.300000000017
Run:  2  --- Steps taken:  93  --- Loss:  10980.200000000008
Run:  34  --- Steps taken:  83  --- Loss:  12766.499999999993
Run:  128  --- Steps taken:  103  --- Loss:  12824.800000000003
Run:  97  --- Steps taken:  102  --- Loss:  12138.200000000012
Run:  3  --- Steps taken:  85  --- Loss:  14262.100000000006
Run:  66  --- Steps taken:  103  --- Loss:  13727.100000000011
Run:  160  --- Steps taken:  74  --- Loss:  7389.600000000001
Run:  35  --- Steps taken:  97  --- Loss:  13195.200000000019
Run:  129  --- Steps taken:  88  --- Loss:  16327.800000000008
Run:  67 

(14988.58460000001, 3863.2864438406355)

# loss_matrix = ((1, 0),(0.2, 0.1)) w/ 100 runs
## max_steps = 100
quarantine_length 5: 15342.928
quarantine_length 5, ct level 2: 14599.878000000002
quarantine_length 10: 15261.373
quarantine_length 5, ct level 1, nt_rate * 10 noticed: 14576.289000000004
quarantine_length inf, ct level 1: 20178.598

## max_steps = 1000 (*)
quarantine_length 5, ct level 1: 14514.140000000009
quarantine_length 10, ct level 1: 15413.901000000007
quarantine_length 5, ct level 2: 14708.250000000007
quarantine_length 5, ct level 3: 15418.860000000004
quarantine_length 2, ct level 1: 15211.508000000003
no quarantines: 14869.428000000004

# loss_matrix = ((1, 0),(0.2, 0.1)) w/ 1000 runs
quarantine_length 5, ct level 1: ~14948 loss and ~3747 std
quarantine_length 10, ct level 1: ~15314 loss and ~3720 std
quarantine_length 5, ct level 2: ~14725 loss and ~3844 std
quarantine_length 10, ct level 2: ~15281 loss and ~4066 std
quarantine_length 5, ct level 3: ~14989 loss and ~3863 std
